drivers/mtd/at25ee.c

```diff
@@ -0,0 +1,1033 @@
+
+      /* Deselect the FLASH */
+
+      SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), false);
+      at25ee_unlock(priv->spi);
+
+      /* Given that writing could take up to a few milliseconds,
+       * the following short delay in the "busy" case will allow
+       * other peripherals to access the SPI bus.
+       */
+
+      if ((status & AT25EE_SR_WIP) != 0)
+        {
+          nxsig_usleep(1000);
+        }
+    }
+  while ((status & AT25EE_SR_WIP) != 0);
+}
+
+/****************************************************************************
+ * Name: at25ee_writeenable
+ *
+ * Description:
+ *    Enable or disable write operations.
+ *    This is required before any write, since a lot of operations
+ *    automatically disable the write latch.
+ *
+ * Input Parameters:
+ *   priv        - a reference to the device structure
+ *   enable      - enable (true) or disable(false) write operations
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void at25ee_writeenable(FAR struct at25ee_dev_s *priv, int enable)
+{
+  at25ee_lock(priv->spi);
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), true);
+
+  SPI_SEND(priv->spi, enable ? AT25EE_CMD_WREN : AT25EE_CMD_WRDIS);
+
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), false);
+  at25ee_unlock(priv->spi);
+}
+
+/****************************************************************************
+ * Name: at25ee_writepage
+ *
+ * Description:
+ *   Write data to the EEPROM, NOT crossing page boundaries.
+ *
+ * Input Parameters:
+ *   priv        - a reference to the device structure
+ *   devaddr     - the address to start the write
+ *   data        - pointer to data buffer to write
+ *   len         - length of the data to write
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void at25ee_writepage(FAR struct at25ee_dev_s *priv, uint32_t devaddr,
+                             FAR const uint8_t *data, size_t len)
+{
+  at25ee_lock(priv->spi);
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), true);
+
+  at25ee_sendcmd(priv->spi, AT25EE_CMD_WRITE, priv->addrlen, devaddr);
+  SPI_SNDBLOCK(priv->spi, data, len);
+
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), false);
+  at25ee_unlock(priv->spi);
+}
+
+/****************************************************************************
+ * Name: at25ee_eraseall
+ *
+ * Description:
+ *   Erase all data in the device
+ *
+ * Input Parameters:
+ *   priv        - a reference to the device structure
+ *   devaddr     - the address to start the write
+ *   data        - pointer to data buffer to write
+ *   len         - length of the data to write
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static int at25ee_eraseall(FAR struct at25ee_dev_s *priv)
+{
+  uint8_t *buf;
+  int startblock = 0;
+
+  DEBUGASSERT(priv);
+
+  buf = kmm_malloc(priv->pgsize);
```

Comment:
is it right to write the random bit to eeprom for erasing

Reply:
@xiaoxiang781216 It's a malloc not memset - have I had a complete brain fail? Or misunderstood/misued kmm_malloc?

Reply:
the buffer returned by kmm_malloc contain the random bits.

Reply:
Missing, accidentally deleted, memcpy. I recall, now, that I added a second buf to read back the erased memory to do a memcmp - like a verify - while I was developing this, so when removing that I must have deleted the memcpy oo. Doh!!